### PR TITLE
Fetch pep

### PR DESCRIPTION
**Pull request by hargoniX**
Adds bot.pep and bot.get_pep for fetching information about a pep
Example

Depends on #76

### CODE REVIEW

**Comment by 1mn** [comment]
#76 is now merged.

**Comment by hargoniX** [comment]
image
This is how it looks like now

File: bot/cogs/utils.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+from bot.constants import Roles
+from bot.decorators import with_role
+
+log = logging.getLogger(__name__)
+
+
+class Utils:
+    """
+    Helpful commands
```
Comment (1mn) [inline bot/cogs/utils.py]: This docstring is fairly useless. All commands are "helpful".

File: bot/cogs/utils.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+        self.base_pep_url = "http://www.python.org/dev/peps/pep-"
+        self.base_github_pep_url = "https://raw.githubusercontent.com/python/peps/master/pep-"
+
+    @command(name="pep()", aliases=["pep", "get_pep"])
+    @with_role(Roles.verified)
+    async def pep_search(self, ctx: Context, pep_number: str):
+        """
+        Fetches information about a PEP and sends it to the user
+        """
```
Comment (1mn) [inline bot/cogs/utils.py]: multiline docstrings should be followed by blank line.

File: bot/cogs/utils.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+            if "Python-Version" in pep_header:
+                pep_embed.add_field(name="Python-Version", value=pep_header["Python-Version"])
+            if "Created" in pep_header:
+                pep_embed.add_field(name="Created", value=pep_header["Created"])
+            if "Type" in pep_header:
+                pep_embed.add_field(name="Type", value=pep_header["Type"])
+
+        elif response.status == 404:
+            log.trace("PEP was not found")
+            not_found = f"PEP {pep_number} is not an existing one"
```
Comment (1mn) [inline bot/cogs/utils.py]: that's.. not how you say that. try `does not exist`. You're so German sometimes.

File: bot/cogs/utils.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+            pep_embed = Embed(title="PEP not found", description=not_found)
```
Comment (1mn) [inline bot/cogs/utils.py]: please format this to be more like the error embeds in the other cogs. It's missing a color, for instance.

File: bot/cogs/utils.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+
+        else:
+            log.trace(f"HTTP error {response.status} during request of PEP")
```
Comment (1mn) [inline bot/cogs/utils.py]: this is also poorly formulated, and could probably contain more useful info. Perhaps the entire response could be logged, as well as more information about what the user requested.

File: bot/cogs/utils.py (hunk 3 above)
Comment (hargoniX) [reply bot/cogs/utils.py]: brb fixing my nationality.

**Comment by gdude2002** [comment]
**In order to expedite our current plans, we're requiring that all PRs be closed or merged.**

You have some oustanding review comments to address!

File: bot/cogs/utils.py (hunk 2 above)
Comment (kingdom5500) [reply bot/cogs/utils.py]: i think you should punctuate your docstring, by the way.

File: bot/cogs/utils.py [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+
+    @command(name="pep()", aliases=["pep", "get_pep"])
+    @with_role(Roles.verified)
+    async def pep_search(self, ctx: Context, pep_number: str):
+        """
+        Fetches information about a PEP and sends it to the user
+        """
+
+        # Attempt to fetch the PEP from Github.
+        pep_url = f"{self.base_github_pep_url}{pep_number.zfill(4)}.txt"
```
Comment (kingdom5500) [inline bot/cogs/utils.py]: you can do zfill with fstrings too. i'd personally do `f"{self.base_github_pep_url}{pep_number:04}.txt"`.

File: bot/cogs/utils.py [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+
+            pep_content = await response.text()
+
+            # Taken from https://github.com/python/peps/blob/master/pep0/pep.py#L179
+            pep_header = HeaderParser().parse(StringIO(pep_content))
+
+            # Assemble the embed
+            pep_embed = Embed(
+                title=f"**PEP {pep_number} - {pep_header['Title']}**",
+                description=f"[Link]({self.base_pep_url+pep_number.zfill(4)})",
```
Comment (kingdom5500) [inline bot/cogs/utils.py]: shouldn't there be spaces surrounding this + operator?   or you could do `f"[Link]({self.base_pep_url}{pep_number:04})"` as i suggested above.

File: bot/cogs/utils.py [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+        elif response.status == 404:
+            log.trace("PEP was not found")
+            not_found = f"PEP {pep_number} does not exist."
+            pep_embed = Embed(title="PEP not found", description=not_found)
+            pep_embed.colour = Colour.red()
+
+        else:
+            log.trace(f"The user requested PEP {pep_number}, ",
+                      "but the response had an unexpected status code {response.status}.\n{response.text}")
+            error_message = "Unexpected HTTP error during PEP search. Please let us know"
```
Comment (kingdom5500) [inline bot/cogs/utils.py]: the last part should be punctuated yes

**Pull request merged**